File: test/ModularPipelines.UnitTests/Attributes/ModuleRegistrationContextTests.cs
// test/ModularPipelines.UnitTests/Attributes/ModuleRegistrationContextTests.cs
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using ModularPipelines.Attributes.Events;
using ModularPipelines.Context;
using ModularPipelines.Engine.Dependencies;
using ModularPipelines.Modules;
using Moq;

namespace ModularPipelines.UnitTests.Attributes;

public class ModuleRegistrationContextTests
{
    private class ModuleA : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
            => Task.FromResult<string?>("A");
    }

    private class ModuleB : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
            => Task.FromResult<string?>("B");
    }

    [Test]
    public async Task ModuleType_ReturnsCorrectType()
    {
        var context = CreateContext(typeof(ModuleA));

        await Assert.That(context.ModuleType).IsEqualTo(typeof(ModuleA));
    }

    [Test]
    public async Task AddDependency_AddsToDependencyRegistry()
    {
        var dependencyRegistry = new ModuleDependencyRegistry();
        var context = CreateContext(typeof(ModuleA), dependencyRegistry: dependencyRegistry);

        context.AddDependency<ModuleB>();

        var dependencies = dependencyRegistry.GetDynamicDependencies(typeof(ModuleA));
        await Assert.That(dependencies).Contains(typeof(ModuleB));
    }

    [Test]
    public async Task IsModuleRegistered_RegisteredModule_ReturnsTrue()
    {
        var registeredModules = new List<Type> { typeof(ModuleA), typeof(ModuleB) };
        var context = CreateContext(typeof(ModuleA), registeredModules: registeredModules);

        await Assert.That(context.IsModuleRegistered<ModuleB>()).IsTrue();
    }

    [Test]
    public async Task IsModuleRegistered_UnregisteredModule_ReturnsFalse()
    {
        var registeredModules = new List<Type> { typeof(ModuleA) };
        var context = CreateContext(typeof(ModuleA), registeredModules: registeredModules);

        await Assert.That(context.IsModuleRegistered<ModuleB>()).IsFalse();
    }

    [Test]
    public async Task SetMetadata_GetMetadata_RoundTrips()
    {
        var metadataRegistry = new ModuleMetadataRegistry();
        var context = CreateContext(typeof(ModuleA), metadataRegistry: metadataRegistry);

        context.SetMetadata("key", "value");

        await Assert.That(context.GetMetadata<string>("key")).IsEqualTo("value");
    }

    private static ModuleRegistrationContext CreateContext(
        Type moduleType,
        IModuleDependencyRegistry? dependencyRegistry = null,
        IModuleMetadataRegistry? metadataRegistry = null,
        IReadOnlyList<Type>? registeredModules = null)
    {
        var configuration = Mock.Of<IConfiguration>();
        var environment = Mock.Of<IHostEnvironment>();
        var services = new ServiceCollection();

        return new ModuleRegistrationContext(
            moduleType,
            moduleType.GetCustomAttributes(true).OfType<Attribute>().ToList(),
            configuration,
            environment,
            registeredModules ?? new List<Type> { moduleType },
            services,
            dependencyRegistry ?? new ModuleDependencyRegistry(),
            metadataRegistry ?? new ModuleMetadataRegistry());
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Attributes/ModuleRegistrationContextTests.fs
namespace ModularPipelines.UnitTests.FSharp.Attributes

open System
open System.Threading
open System.Threading.Tasks
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.Options
open ModularPipelines.Attributes.Events
open ModularPipelines.Context
open ModularPipelines.Engine.Dependencies
open ModularPipelines.Modules
open ModularPipelines.Options
open Moq
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module ModuleRegistrationContextTests =
    type ModuleA() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            Task.FromResult<string>("A")

    type ModuleB() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            Task.FromResult<string>("B")

    let createContext
            (moduleType: Type)
            (dependencyRegistry: IModuleDependencyRegistry option)
            (metadataRegistry: IModuleMetadataRegistry option)
            (registeredModules: Type list option) =
        let configuration = Mock.Of<IConfiguration>()
        let environment = Mock.Of<IHostEnvironment>()
        let services = ServiceCollection()
        let depReg = defaultArg dependencyRegistry (ModuleDependencyRegistry() :> IModuleDependencyRegistry)
        let metaReg = defaultArg metadataRegistry (ModuleMetadataRegistry(Options.Create(ModuleRegistrationOptions())) :> IModuleMetadataRegistry)
        let regModules = defaultArg registeredModules [ moduleType ] |> List.toArray :> System.Collections.Generic.IReadOnlyList<Type>
        ModuleRegistrationContext(
            moduleType,
            moduleType.GetCustomAttributes(true) |> Array.choose (function :? Attribute as a -> Some a | _ -> None) |> Array.toList :> System.Collections.Generic.IReadOnlyList<Attribute>,
            configuration,
            environment,
            regModules,
            services,
            depReg,
            metaReg)

type ModuleRegistrationContextTests() =
    [<Test>]
    member _.ModuleType_ReturnsCorrectType() = async {
        let context = ModuleRegistrationContextTests.createContext typeof<ModuleRegistrationContextTests.ModuleA> None None None
        do! check(Assert.That(context.ModuleType).IsEqualTo(typeof<ModuleRegistrationContextTests.ModuleA>))
    }

    [<Test>]
    member _.AddDependency_AddsToDependencyRegistry() = async {
        let dependencyRegistry = ModuleDependencyRegistry()
        let context = ModuleRegistrationContextTests.createContext typeof<ModuleRegistrationContextTests.ModuleA> (Some (dependencyRegistry :> IModuleDependencyRegistry)) None None
        context.AddDependency<ModuleRegistrationContextTests.ModuleB>()
        let dependencies = dependencyRegistry.GetDynamicDependencies(typeof<ModuleRegistrationContextTests.ModuleA>)
        do! check(Assert.That(dependencies |> Seq.contains typeof<ModuleRegistrationContextTests.ModuleB>).IsTrue())
    }

    [<Test>]
    member _.IsModuleRegistered_RegisteredModule_ReturnsTrue() = async {
        let registeredModules = [ typeof<ModuleRegistrationContextTests.ModuleA>; typeof<ModuleRegistrationContextTests.ModuleB> ]
        let context = ModuleRegistrationContextTests.createContext typeof<ModuleRegistrationContextTests.ModuleA> None None (Some registeredModules)
        do! check(Assert.That(context.IsModuleRegistered<ModuleRegistrationContextTests.ModuleB>()).IsTrue())
    }

    [<Test>]
    member _.IsModuleRegistered_UnregisteredModule_ReturnsFalse() = async {
        let registeredModules = [ typeof<ModuleRegistrationContextTests.ModuleA> ]
        let context = ModuleRegistrationContextTests.createContext typeof<ModuleRegistrationContextTests.ModuleA> None None (Some registeredModules)
        do! check(Assert.That(context.IsModuleRegistered<ModuleRegistrationContextTests.ModuleB>()).IsFalse())
    }

    [<Test>]
    member _.SetMetadata_GetMetadata_RoundTrips() = async {
        let metadataRegistry = ModuleMetadataRegistry(Options.Create(ModuleRegistrationOptions()))
        let context = ModuleRegistrationContextTests.createContext typeof<ModuleRegistrationContextTests.ModuleA> None (Some (metadataRegistry :> IModuleMetadataRegistry)) None
        context.SetMetadata("key", "value")
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(context.GetMetadata<string>("key")), "value"))
    }
